File: taskmanager/src/test/java/com/example/alexeyglushkov/taskmanager/task/PriorityTaskProviderTest.java
package com.example.alexeyglushkov.taskmanager.task;

import android.os.Handler;
import android.os.Looper;
import androidx.annotation.NonNull;

import org.junit.Before;
import org.junit.Test;

import static junit.framework.Assert.assertEquals;

import org.mockito.Mockito;

import java.util.Arrays;

/**
 * Created by alexeyglushkov on 09.08.15.
 */

public class PriorityTaskProviderTest {
    private TaskPoolTestSet poolTestSet;
    private TaskProviderTestSet providerTest;
    protected TaskProvider taskProvider;

    @Before
    public void setUp() throws Exception {
        taskProvider = prepareTaskProvider();

        poolTestSet = new TaskPoolTestSet();
        providerTest = new TaskProviderTestSet();

        poolTestSet.before(taskProvider);
        providerTest.before(taskProvider);
    }

    @NonNull
    protected TaskProvider prepareTaskProvider() {
        return new PriorityTaskProvider(new Handler(Looper.myLooper()), "TestId");
    }

    protected PriorityTaskProvider getPriorityTaskProvider() {
        return (PriorityTaskProvider)taskProvider;
    }

    // PriorityTaskProviderTests

    @Test
    public void testUpdatePriorities() {
        // Arrange
        TaskProvider.Listener listener = Mockito.mock(TaskProvider.Listener.class);

        taskProvider.addListener(listener);
        taskProvider.addTask(TestTasks.createTestTaskSpy("a", 1, 1));
        taskProvider.addTask(TestTasks.createTestTaskSpy("b", 2, 2));
        taskProvider.addTask(TestTasks.createTestTaskSpy("c", 3, 3));
        taskProvider.addTask(TestTasks.createTestTaskSpy("d", 2, 4));
        taskProvider.addTask(TestTasks.createTestTaskSpy("e", 1, 5));
        taskProvider.addTask(TestTasks.createTestTaskSpy("f", 3, 6));

        // Act
        getPriorityTaskProvider().updatePriorities(new PriorityTaskProvider.PriorityProvider() {
            @Override
            public int getPriority(Task task) {
                if (task.getTaskType() == 2) {
                    return 2 * task.getTaskPriority();
                }
                return task.getTaskPriority();
            }
        });

        Task task = taskProvider.getTopTask(null);

        // Verify
        assertEquals("d", task.getTaskId());
    }

    @Test
    public void testTopTaskWithPriorityWithoutFilter() {
        // Arrange
        taskProvider.addTask(TestTasks.createTestTaskSpy("a", 1, 1));
        taskProvider.addTask(TestTasks.createTestTaskSpy("b", 2, 2));
        taskProvider.addTask(TestTasks.createTestTaskSpy("c", 1, 3));
        taskProvider.addTask(TestTasks.createTestTaskSpy("d", 2, 4));
        taskProvider.addTask(TestTasks.createTestTaskSpy("e", 1, 5));
        taskProvider.addTask(TestTasks.createTestTaskSpy("f", 2, 6));

        // Act
        Task task = taskProvider.getTopTask(null);

        // Verify
        assertEquals("f", task.getTaskId());
    }

    @Test
    public void testGetTopTaskWithPriorityWithFilter() {
        // Arrange
        taskProvider.addTask(TestTasks.createTestTaskSpy("a", 1, 1));
        taskProvider.addTask(TestTasks.createTestTaskSpy("b", 2, 2));
        taskProvider.addTask(TestTasks.createTestTaskSpy("c", 3, 3));
        taskProvider.addTask(TestTasks.createTestTaskSpy("d", 2, 4));
        taskProvider.addTask(TestTasks.createTestTaskSpy("e", 1, 5));
        taskProvider.addTask(TestTasks.createTestTaskSpy("f", 3, 6));

        // Act
        Task task = taskProvider.getTopTask(Arrays.asList(new Integer[]{3}));

        // Verify
        assertEquals("e", task.getTaskId());
    }

    // ProviderTests

    @Test
    public void testGetTopTaskWithBlockedTask() {
        providerTest.getTopTaskWithBlockedTask();
    }

    @Test
    public void testTakeTopTaskWithBlockedTask() {
        providerTest.takeTopTaskWithBlockedTask();
    }

    @Test
    public void testSetProviderId() {
        providerTest.setProviderId();
    }

    @Test
    public void testSetPriority() {
        providerTest.setPriority();
    }

    @Test
    public void testGetTopTaskWithoutFilter() {
        providerTest.getTopTaskWithoutFilter();
    }

    @Test
    public void testGetTopTaskWithFilter() {
        providerTest.getTopTaskWithFilter();
    }

    @Test
    public void testTakeTopTaskWithFilter() {
        providerTest.takeTopTaskWithFilter();
    }

    @Test
    public void testRemoveTaskWithUnknownType() {
        providerTest.removeTaskWithUnknownType();
    }

    // PoolTests

    @Test
    public void testSetGetHandler() {
        poolTestSet.setGetHandler();
    }

    @Test
    public void testAddTask() {
        poolTestSet.addTask();
    }

    @Test
    public void testAddTaskAddStatusListener() {
        poolTestSet.addTaskAddStatusListener();
    }

    @Test
    public void testAddStartedTask() {
        poolTestSet.addStartedTask();
    }

    @Test
    public void testRemoveTask() {
        poolTestSet.removeTask();
    }

    @Test
    public void testRemoveUnknownTask() {
        poolTestSet.removeUnknownTask();
    }

    @Test
    public void testGetTask() {
        poolTestSet.getTask();
    }

    @Test
    public void testGetUnknownTask() {
        poolTestSet.getUnknownTask();
    }

    @Test
    public void testGetTaskCount() {
        poolTestSet.getTaskCount();
    }

    @Test
    public void testGetTaskCount2() {
        poolTestSet.getTaskCount2();
    }

    @Test
    public void testSetGetUserData() {
        poolTestSet.setGetUserData();
    }

    @Test
    public void testAddStateListener() {
        poolTestSet.addStateListener();
    }

    @Test
    public void testRemoveStateListener() {
        poolTestSet.removeStateListener();
    }

    @Test
    public void testChangeTaskStatus() {
        poolTestSet.changeTaskStatus();
    }

    @Test
    public void testCheckTaskRemovingAfterFinishing() {
        poolTestSet.checkTaskRemovingAfterFinishing();
    }
}

File: taskmanager/src/test/java/com/example/alexeyglushkov/taskmanager/task/PriorityTaskProviderTest.kt
package com.example.alexeyglushkov.taskmanager.task

import com.nhaarman.mockitokotlin2.mock

import org.junit.Before
import org.junit.Test
import org.junit.Assert.*

/**
 * Created by alexeyglushkov on 09.08.15.
 */

open class PriorityTaskProviderTest {
    private lateinit var poolTestSet: TaskPoolTestSet
    private lateinit var providerTest: TaskProviderTestSet
    protected lateinit var taskProvider: TaskProvider

    protected open val priorityTaskProvider: PriorityTaskProvider
        get() = taskProvider as PriorityTaskProvider

    @Before
    @Throws(Exception::class)
    fun setUp() {
        taskProvider = prepareTaskProvider()

        poolTestSet = TaskPoolTestSet()
        providerTest = TaskProviderTestSet()

        poolTestSet.before(taskProvider)
        providerTest.before(taskProvider)
    }

    protected open fun prepareTaskProvider(): TaskProvider {
        return PriorityTaskProvider(TestTasks.createTestScopeThreadRunner(), "TestId")
    }

    // PriorityTaskProviderTests

    @Test
    fun testUpdatePriorities() {
        // Arrange
        val listener = mock<TaskPool.Listener>()

        taskProvider.addListener(listener)
        taskProvider.addTask(TestTasks.createTestTaskSpy("a", 1, 1))
        taskProvider.addTask(TestTasks.createTestTaskSpy("b", 2, 2))
        taskProvider.addTask(TestTasks.createTestTaskSpy("c", 3, 3))
        taskProvider.addTask(TestTasks.createTestTaskSpy("d", 2, 4))
        taskProvider.addTask(TestTasks.createTestTaskSpy("e", 1, 5))
        taskProvider.addTask(TestTasks.createTestTaskSpy("f", 3, 6))

        // Act
        priorityTaskProvider.updatePriorities(object : PriorityTaskProvider.PriorityProvider {
            override fun getPriority(task: Task): Int {
                return if (task.taskType == 2) {
                    2 * task.taskPriority
                } else task.taskPriority
            }
        })

        val task = taskProvider.getTopTask()

        // Verify
        assertEquals("d", task!!.taskId)
    }

    @Test
    fun testTopTaskWithPriorityWithoutFilter() {
        // Arrange
        taskProvider.addTask(TestTasks.createTestTaskSpy("a", 1, 1))
        taskProvider.addTask(TestTasks.createTestTaskSpy("b", 2, 2))
        taskProvider.addTask(TestTasks.createTestTaskSpy("c", 1, 3))
        taskProvider.addTask(TestTasks.createTestTaskSpy("d", 2, 4))
        taskProvider.addTask(TestTasks.createTestTaskSpy("e", 1, 5))
        taskProvider.addTask(TestTasks.createTestTaskSpy("f", 2, 6))

        // Act
        val task = taskProvider.getTopTask()

        // Verify
        assertEquals("f", task!!.taskId)
    }

    @Test
    fun testGetTopTaskWithPriorityWithFilter() {
        // Arrange
        taskProvider.addTask(TestTasks.createTestTaskSpy("a", 1, 1))
        taskProvider.addTask(TestTasks.createTestTaskSpy("b", 2, 2))
        taskProvider.addTask(TestTasks.createTestTaskSpy("c", 3, 3))
        taskProvider.addTask(TestTasks.createTestTaskSpy("d", 2, 4))
        taskProvider.addTask(TestTasks.createTestTaskSpy("e", 1, 5))
        taskProvider.addTask(TestTasks.createTestTaskSpy("f", 3, 6))

        taskProvider.taskFilter = object : TaskProvider.TaskFilter {
            override fun getFilteredTaskTypes(): List<Int> {
                return listOf(3)
            }
        }

        // Act
        val task = taskProvider.getTopTask()

        // Verify
        assertEquals("e", task!!.taskId)
    }

    // ProviderTests

    @Test
    fun getTopTaskWithDependantTask() {
        providerTest.getTopTaskWithDependantTask()
    }

    @Test
    fun getTopTaskWithBlockedTask() {
        providerTest.getTopTaskWithBlockedTask()
    }

    @Test
    fun takeTopTaskWithDependantTask() {
        providerTest.takeTopTaskWithDependantTask()
    }

    @Test
    fun takeTopTaskWithBlockedTask() {
        providerTest.takeTopTaskWithBlockedTask()
    }

    @Test
    fun testSetProviderId() {
        providerTest.setProviderId()
    }

    @Test
    fun testSetPriority() {
        providerTest.setPriority()
    }

    @Test
    fun testGetTopTaskWithoutFilter() {
        providerTest.getTopTaskWithoutFilter()
    }

    @Test
    fun testGetTopTaskWithFilter() {
        providerTest.getTopTaskWithFilter()
    }

    @Test
    fun testTakeTopTaskWithFilter() {
        providerTest.takeTopTaskWithFilter()
    }

    @Test
    fun testRemoveTaskWithUnknownType() {
        providerTest.removeTaskWithUnknownType()
    }

    // PoolTests

    @Test
    fun testSetGetHandler() {
        poolTestSet.setGetHandler()
    }

    @Test
    fun testAddTask() {
        poolTestSet.addTask()
    }

    @Test
    fun addTaskCallsAddStatusListener() {
        poolTestSet.addTaskCallsAddStatusListener()
    }

    @Test
    fun removeTaskCallsRemoveTaskStatusListener() {
        poolTestSet.removeTaskCallsRemoveTaskStatusListener()
    }

    @Test
    fun testAddStartedTask() {
        poolTestSet.addStartedTask()
    }

    @Test
    fun testRemoveTask() {
        poolTestSet.removeTask()
    }

    @Test
    fun testRemoveUnknownTask() {
        poolTestSet.removeUnknownTask()
    }

    @Test
    fun testGetTask() {
        poolTestSet.getTask()
    }

    @Test
    fun testGetUnknownTask() {
        poolTestSet.getUnknownTask()
    }

    @Test
    fun testGetTaskCount() {
        poolTestSet.getTaskCount()
    }

    @Test
    fun testGetTaskCount2() {
        poolTestSet.getTaskCount2()
    }

    @Test
    fun testSetGetUserData() {
        poolTestSet.setGetUserData()
    }

    @Test
    fun testAddStateListener() {
        poolTestSet.addStateListener()
    }

    @Test
    fun testRemoveStateListener() {
        poolTestSet.removeStateListener()
    }

    @Test
    fun testChangeTaskStatus() {
        poolTestSet.changeTaskStatus()
    }

    @Test
    fun testCheckTaskRemovingAfterFinishing() {
        poolTestSet.checkTaskRemovingAfterFinishing()
    }

    @Test
    fun testCancelTask() {
        poolTestSet.cancelTask()
    }
}
